feat: add nydusify commit command

PerseidMeteor opened the pull request:
## Relevant Issue (if applicable)
[support nydus commit operation](https://github.com/dragonflyoss/nydus/issues/1307)

## Details
_Please describe the details of PullRequest._

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

PerseidMeteor commented on contrib/nydusify/pkg/commiter/util.go:
```diff
@@ -0,0 +1,25 @@
+package commiter
+
+import (
+	// "context"
```
forget, remove later

imeoer commented:
Thanks for the implementation! Please also add a test case in https://github.com/dragonflyoss/nydus/tree/master/smoke/tests (maybe create a `commit_test.go` ?).

PerseidMeteor commented:
> Thanks for the implementation! Please also add a test case in https://github.com/dragonflyoss/nydus/tree/master/smoke/tests (maybe create a `commit_test.go` ?).

I am sorry for I was too late because of my Spring Festival holiday. Commit smoke test was added.
Have a nice day!

davidstack commented:
very nice pr

Desiki-high commented:
Thanks for your pr, please fix the `golang-lint`.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,131 @@
+
+func (c *CommitTestSuite) TestCommitContainer() {
+	ctx := tool.DefaultContext(c.t)
+	c.snapshotter = os.Getenv("SNAPSHOTTER")
```
We'd better use `scenarios := tool.DescartesIterator{}` to manage parameters. :)

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	// choose test mode
+	mode := os.Getenv("COMMIT_MODE")
```
Ditto, should use `tool.DescartesIterator`.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,131 @@
+	tool.RunWithoutOutput(c.t, commitCmd)
+
+	// run committed container
+	_ = tool.RunContainer(c.t, c.testCommittedImage, c.snapshotter, c.testCommittedContainerName)
```
Should also do `nerdctl cp %s:/root/commit commit` to check the file content. :)

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,131 @@
+		return convertMetirc["TargetImageSize"], convertMetirc["ConversionElapsed"]
+	}
+	c.testImage = source
+	return convertMetirc["SourceImageSize"], 0
```
Why we need to consider about metric on this test?

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+func TestCommit(t *testing.T) {
+	if os.Getenv("COMMIT_TEST") == "" {
```
In what scenario should this test be skipped?

imeoer commented on contrib/nydusify/pkg/utils/constant.go:
```diff
@@ -21,4 +21,7 @@ const (
 	LayerAnnotationUncompressed = "containerd.io/uncompressed"
+
+	LayerAnnotationNydusCommitBlobs = "containerd.io/snapshot/nydus-commit-blobs"
+	LayerAnnotationNydusBlobIDs     = "containerd.io/snapshot/nydus-blob-ids"
```
The constant seems not to be used.

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -1103,6 +1104,94 @@ func main() {
+					EnvVars: []string{"NYDUS_IMAGE"},
+				},
+				&cli.StringFlag{
+					Name:    "containerd-address",
```
Let's we follow the option help message of nerdctl:

`containerd address, optionally with "unix://" prefix [$CONTAINERD_ADDRESS] (default "/run/containerd/containerd.sock")`.

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -1103,6 +1104,94 @@ func main() {
+					Name:     "with-path",
+					Aliases:  []string{"with-mount-path"},
+					Required: false,
+					Usage:    "The directory that need to be committed",
```
The directory in container that need to be committed

imeoer commented on contrib/nydusify/pkg/commiter/commiter.go:
```diff
@@ -0,0 +1,671 @@
+}
+
+func (cm *Commiter) Commit(ctx context.Context, opt Opt) error {
+	targetRef, err := AppendNydusSuffix(opt.TargetRef)
```
Right, should use target reference directly.

imeoer replied:
Let's remove these suffix logics.

imeoer commented on contrib/nydusify/pkg/commiter/nsenter.go:
```diff
@@ -0,0 +1,183 @@
+package commiter
```
Should put the copyright, the codes ported from `https://github.com/Devatoria/go-nsenter`.

davidstack commented:
@imeoer  which release will contain this pr？ we are waiting for this pr for a long time. thanks

imeoer commented:
> @imeoer which release will contain this pr？ we are waiting for this pr for a long time. thanks

Thanks for waiting, it should be included in the next release, v2.3.0.

imeoer commented:
@PerseidMeteor Please also update the docs in https://github.com/dragonflyoss/nydus/blob/master/docs/nydusify.md, better be a workflow like `nydusify convert -> nerdctl run -> nydusify commit -> nydusify run`.

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -1103,6 +1104,108 @@ func main() {
+				}
+
+				withPaths, withoutPaths := parsePaths(c.StringSlice("with-path"))
+				opt := commiter.Opt{
```
Sorry for the misleading, it should be `committer`.

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -1103,6 +1104,108 @@ func main() {
+					WithPaths:         withPaths,
+					WithoutPaths:      withoutPaths,
+				}
+				cm, err := commiter.NewCommiter(opt)
```
Ditto.

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -1103,6 +1104,108 @@ func main() {
+				if err != nil {
+					return errors.Wrap(err, "create commiter")
+				}
+				return cm.Commit(context.Background(), opt)
```
`context.Background()` -> `c.Context()`.

imeoer commented on contrib/nydusify/pkg/commiter/diff/diff.go:
```diff
@@ -0,0 +1,113 @@
+// Ported from buildkit project, copyright The buildkit Authors.
+// https://github.com/moby/buildkit
```
Leave a blank line here.

imeoer commented on contrib/nydusify/pkg/commiter/manager.go:
```diff
@@ -0,0 +1,130 @@
+package commiter
```
Should add nydus developers copyright here.

imeoer commented on contrib/nydusify/pkg/commiter/nsenter.go:
```diff
@@ -0,0 +1,185 @@
+// Ported from go-nsenter project, copyright The go-nsenter Authors.
+// https://github.com/Devatoria/go-nsenter
```
Ditto, leave a blank line here.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,123 @@
+package tests
```
Ditto, add copyright here.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,123 @@
+	checkCmd := fmt.Sprintf("sudo nerdctl cp %s:/root/commit container_commit", committedContainerID)
+	tool.RunWithoutOutput(c.t, checkCmd)
+
+	file, err := os.Open("./container_commit")
```
We'd better put all temp files into the `ctx.Env.WorkDir` directory.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,123 @@
+	tool.RunWithoutOutput(c.t, checkCmd)
+
+	file, err := os.Open("./container_commit")
+	if err != nil {
```
Should use `require.NoError(t, err)`.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,123 @@
+	defer os.Remove(file.Name())
+
+	content, err := io.ReadAll(file)
+	if err != nil {
+		c.t.Fatalf("can not read commit file: %s", err)
+	}
+	fileContent := strings.Trim(string(content), "\n")
```
Ditto, should use `require.NoError(t, err)`.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,123 @@
+	committedContainerID := strings.Trim(tool.RunWithOutput(runCommittedContainerCmd), "\n")
+
+	// check committed file content
+	checkCmd := fmt.Sprintf("sudo nerdctl cp %s:/root/commit container_commit", committedContainerID)
```
May be implement a func like:

```
func checkFileContent(t *testing.T, containerName, path, content string) {
	nerdctlExec(t, containerName, fmt.Sprintf("stat %s && grep -Fxq '%s' %s", path, content, path))
}

func nerdctlExec(t *testing.T, containerName, cmd string) {
	tool.Run(t, fmt.Sprintf("nerdctl exec %s sh -c \"%s\"", containerName, cmd))
}
```

To check the file content consistency.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,123 @@
+		c.t.Fatalf("can not read commit file: %s", err)
+	}
+	fileContent := strings.Trim(string(content), "\n")
+	if fileContent != "This is Nydus commit" {
```
Ditto, `require.Equal`.

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,123 @@
+	// run nydus contaienr
+	containerName := uuid.NewString()
+	runContainerCmd := fmt.Sprintf("sudo nerdctl --snapshotter nydus run  -d -t --insecure-registry --name=%s %s sh", containerName, image)
+	containerID := strings.Trim(tool.RunWithOutput(runContainerCmd), "\n")
```
`defer tool.ClearContainer(c.t, image, "nydus", containerName)`

imeoer commented on smoke/tests/commit_test.go:
```diff
@@ -0,0 +1,123 @@
+	tool.RunWithoutOutput(c.t, commitCmd)
+	// run committed container
+	runCommittedContainerCmd := fmt.Sprintf("sudo nerdctl --snapshotter nydus run  -d -t --insecure-registry --name=%s %s sh", committedContainerName, commmitedImage)
+	committedContainerID := strings.Trim(tool.RunWithOutput(runCommittedContainerCmd), "\n")
```
`defer tool.ClearContainer(c.t, commmitedImage, "nydus", committedContainerName)`

imeoer commented:
![image](https://github.com/dragonflyoss/nydus/assets/1524576/e163731e-7d4d-4db3-8c08-446ef4334126)

It seems the CI is broken.

PerseidMeteor committed:
feat: ported write overlay upperdir capability
ported  capability of get and write diff between overlayfs upper and lower.

Signed-off-by: YuQiang <y_q_email@163.com>

PerseidMeteor committed:
feat: add nydusify commit command
add nydusify commit command to commit a nydus container into nydus image

Signed-off-by: YuQiang <y_q_email@163.com>

PerseidMeteor commented:
> ![image](https://private-user-images.githubusercontent.com/1524576/311563239-e163731e-7d4d-4db3-8c08-446ef4334126.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMTU0NTMsIm5iZiI6MTcxMDMxNTE1MywicGF0aCI6Ii8xNTI0NTc2LzMxMTU2MzIzOS1lMTYzNzMxZS03ZDRkLTRkYjMtOGMwOC00NDZlZjQzMzQxMjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTNUMDczMjMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2U2YzU1MWNiZmRiN2Q3OTQwZTZhMDAyMjA3ZDRkYjEwZjBjYTdlNjlkYjdmNmU1NjE0Yzg1MDUzZjhhZGUwOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Im1CTx4_84n0RlAmlFL9CYLes-ZT-nFD3TDE29ggdkg)
> 
> It seems the CI broken.

It made me confused. Is there any idea to slove this problem?

PerseidMeteor committed:
feat: add smoke test for nydusify commit

Signed-off-by: YuQiang <y_q_email@163.com>

PerseidMeteor committed:
action: merge and move prepare.sh
remove misc/performance/prepare.sh and misc/performance/prepare.sh and merge to misc/prepare.sh

Signed-off-by: YuQiang <y_q_email@163.com>
